### PR TITLE
Perfect Jacobian for large deformation J2 plasticity with nonlinear hardening

### PR DESCRIPTION
**Pull request by hugary1995**
## Reason

We don't currently have a non-AD version plasticity stress update (with nonlinear hardening) with perfect Jacobian.

## Design

Derive and implement the perfect Jacobian in the new mechanics system.

## Impact

This will probably make @dschwen unhappy as this makes the advantage of AD diminishing for single physics mechanic simulations... But AD is always helpful in a multiphysics setting. @recuero might also be interested.

The Jacobian derivation is attached here:
[SimoHughesJ2Jacobian.pdf](https://github.com/idaholab/moose/files/8919779/SimoHughesJ2Jacobian.pdf)
which you could follow to derive the perfect Jacobian for anisotropic plasticity.

### CODE REVIEW

**Comment by hugary1995** [comment]
Uhh I am basing this PR on #21227. I'll rebase once that one is merged.

**Comment by hugary1995** [comment]
@dschwen This is what I meant by needing the second derivative of the hardening function. See equations 19-21 for example.

File: modules/tensor_mechanics/include/materials/lagrangian/ComputeSimoHughesJ2PlasticityStress.h [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "ComputeLagrangianStressPK1.h"
+#include "GuaranteeConsumer.h"
+#include "ElasticityTensorTools.h"
```
Comment (reverendbedford) [inline modules/tensor_mechanics/include/materials/lagrangian/ComputeSimoHughesJ2PlasticityStress.h:14]: I may find out below, but do you need this include?

File: modules/tensor_mechanics/include/materials/lagrangian/ComputeSimoHughesJ2PlasticityStress.h [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+  /// @{ The return mapping residual and derivative
+  virtual Real computeReferenceResidual(const Real & effective_trial_stress,
+                                        const Real & scalar) override;
+  virtual Real computeResidual(const Real & effective_trial_stress, const Real & scalar) override;
+  virtual Real computeDerivative(const Real & effective_trial_stress, const Real & scalar) override;
+  virtual void
+  preStep(const Real & scalar_old, const Real & residual, const Real & jacobian) override;
+  /// @}
+
+  const MaterialPropertyName _elasticity_tensor_name;
```
Comment (reverendbedford) [inline modules/tensor_mechanics/include/materials/lagrangian/ComputeSimoHughesJ2PlasticityStress.h:46]: I feel like we shouldn't need this in the end

File: modules/tensor_mechanics/src/kernels/lagrangian/TotalLagrangianStressDivergence.C [hunk 3]
```diff
@@ -124,7 +123,9 @@ TotalLagrangianStressDivergence::eigenstrainJacobianComponent(unsigned int cvar,
     B = _def_grad[_qp];
   }
 
-  RankFourTensor U = 0.5 * (A.mixedProductIkJl(B.transpose()) + A.mixedProductIlJk(B.transpose()));
```
Comment (reverendbedford) [inline modules/tensor_mechanics/src/kernels/lagrangian/TotalLagrangianStressDivergence.C:127]: This probably belongs in a separate PR, but I won't object.

Edit: or maybe this is from one of the others...

File: modules/tensor_mechanics/src/materials/lagrangian/ComputeSimoHughesJ2PlasticityStress.C [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+  ComputeLagrangianStressPK1::initQpStatefulProperties();
+  _be[_qp].setToIdentity();
+  _ep[_qp] = 0;
+}
+
+void
+ComputeSimoHughesJ2PlasticityStress::computeQpPK1Stress()
+{
+  usingTensorIndices(i, j, k, l, m);
+  const Real G = ElasticityTensorTools::getIsotropicShearModulus(_elasticity_tensor[_qp]);
```
Comment (reverendbedford) [inline modules/tensor_mechanics/src/materials/lagrangian/ComputeSimoHughesJ2PlasticityStress.C:75]: Oh, okay.  I guess this is fine for the way things are setup now.  If we need to add modularity on the elastic model we can do so later.  Ignore my elasticity tensor comments from above.

**Review by reverendbedford — DISMISSED** [review]
I tried to just review the "actual" new content in this PR versus the chain of related prerequisite PRs.  If I sorted everything out correctly I don't think I see any need for any changes.

**Comment by dschwen** [comment]
> This will probably make @dschwen unhappy as this makes the advantage of AD diminishing for single physics mechanic simulations...

🤣

File: framework/include/utils/RankFourTensor.h [hunk 5]
```diff
@@ -757,3 +753,20 @@ RankFourTensorTempl<T>::inverse() const
 
   return result;
 }
+
+template <typename T>
+template <int m>
+RankThreeTensorTempl<T>
+RankFourTensorTempl<T>::contraction(const VectorValue<T> & b) const
+{
+  RankThreeTensorTempl<T> result;
```
Comment (dschwen) [inline framework/include/utils/RankFourTensor.h:762]: Ok, I'll fix up my PR first.

File: modules/tensor_mechanics/include/materials/lagrangian/ComputeSimoHughesJ2PlasticityStress.h (hunk 1 above)
Comment (hugary1995) [reply modules/tensor_mechanics/include/materials/lagrangian/ComputeSimoHughesJ2PlasticityStress.h:14]: Yes.

Comment (hugary1995) [reply modules/tensor_mechanics/include/materials/lagrangian/ComputeSimoHughesJ2PlasticityStress.h:14]: This is for extracting the isotropic parameters.

File: modules/tensor_mechanics/doc/content/source/materials/lagrangian/ComputeSimoHughesJ2PlasticityStress.md [hunk 6]
```diff
@@ -0,0 +1,87 @@
+# ComputeSimoHughesJ2PlasticityStress
+
+!syntax description /Materials/ComputeSimoHughesJ2PlasticityStress
```
Comment (recuero) [inline modules/tensor_mechanics/doc/content/source/materials/lagrangian/ComputeSimoHughesJ2PlasticityStress.md:3]: Can you add a reference to the paper (I assume there is a paper besides the book)

File: modules/tensor_mechanics/include/materials/SingleVariableReturnMappingSolution.h [hunk 7]
```diff
@@ -112,6 +112,12 @@ class SingleVariableReturnMappingSolutionTempl
   virtual Real computeReferenceResidual(const GenericReal<is_ad> & effective_trial_stress,
                                         const GenericReal<is_ad> & scalar) = 0;
```
Comment (recuero) [inline modules/tensor_mechanics/include/materials/SingleVariableReturnMappingSolution.h:114]: Maybe good to document here the use case

File: modules/tensor_mechanics/src/materials/lagrangian/ComputeSimoHughesJ2PlasticityStress.C [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+                                                         : std::sqrt(3. / 2.) * s / s.norm();
+  Real s_eff = s.doubleContraction(_Np[_qp]);
+
+  // Compute the derivative of the strain before return mapping
+  if (_fe_problem.currentlyComputingJacobian())
+    _d_be_d_F = _F_old[_qp].inverse().times<l, m, i, j, k, m>(
+        (I.times<i, k, j, l>(f_bar * _be_old[_qp].transpose()) +
+         I.times<j, k, i, l>(f_bar * _be_old[_qp])) /
+            std::cbrt(f.det()) -
+        2. / 3. * _be[_qp].times<i, j, l, k>(_inv_df[_qp]));
```
Comment (recuero) [inline modules/tensor_mechanics/src/materials/lagrangian/ComputeSimoHughesJ2PlasticityStress.C:98]: These loops and others are hard to follow, probably because they have to.  Is there any algorithm in the literature you can point to in a comment?

Comment (hugary1995) [reply modules/tensor_mechanics/src/materials/lagrangian/ComputeSimoHughesJ2PlasticityStress.C:98]: Sorry which loops are you referring to? If you mean the overall algorithm, then I have included a writeup in the PR description. I think there is also a paper by Simo which has this algorithm. I'll find it and cite it in the documentation.

File: modules/tensor_mechanics/include/materials/SingleVariableReturnMappingSolution.h (hunk 7 above)
Comment (hugary1995) [reply modules/tensor_mechanics/include/materials/SingleVariableReturnMappingSolution.h:114]: 👍

File: modules/tensor_mechanics/src/materials/lagrangian/ComputeSimoHughesJ2PlasticityStress.C (hunk 8 above)
Comment (recuero) [reply modules/tensor_mechanics/src/materials/lagrangian/ComputeSimoHughesJ2PlasticityStress.C:98]: I ended up not selecting any loop... Yeah, I was referring to the overall algorithm. My bad, I missed that document you attached to the description: Can that document be md-fied? I am afraid we may not come across it when we or users most need it.

File: modules/tensor_mechanics/doc/content/source/materials/lagrangian/ComputeSimoHughesJ2PlasticityStress.md (hunk 6 above)
Comment (hugary1995) [reply modules/tensor_mechanics/doc/content/source/materials/lagrangian/ComputeSimoHughesJ2PlasticityStress.md:3]: Added

File: modules/tensor_mechanics/src/materials/lagrangian/ComputeSimoHughesJ2PlasticityStress.C (hunk 8 above)
Comment (hugary1995) [reply modules/tensor_mechanics/src/materials/lagrangian/ComputeSimoHughesJ2PlasticityStress.C:98]: Yep, the algorithm is included in the md file.

**Review by dschwen — APPROVED** [review]
